### PR TITLE
feat(versioning): add regex versioner

### LINKED ISSUE #4264: Allow user-defined version schemes
Note: I am already working on a solution to this feature request. ~Will convert this to a PR Soon™ after getting some comments and cleaning up my branch~ PR is available at #4273. I would mostly hope to use this issue as a place to RFC my proposed solution, in case the maintainers would be interested in an alternate approach or there's anything I'm unaware of.

**What would you like Renovate to be able to do?**
I've been recently working through [this issue](https://github.com/renovatebot/config-help/issues/319), which amounts to "the [Python docker image tags versions somewhat uniquely](https://hub.docker.com/_/python?tab=tags), and Renovate can't quite support that". [Other similar issues](https://github.com/renovatebot/renovate/issues/2438) have been raised before. Support for these slightly weird/unique version schemes would be fantastic.

**Describe the solution you'd like**
I propose the addition of a configurable `versionScheme`, which allows users to specify the format of any given dependency in order to plug these gaps, perhaps with an optional `versionConfig`

**Describe alternatives you've considered**
There aren't many other options here; most of these repos have some sort of almost-right fallback, for example:
- the python docker images can use the `docker` versioner, but then `ignoreUnstable` will break
- the python docker images can use the `pep440` versioner, but then versioning will break entirely if you use the various `-alpine` / `-slim` flags

**Additional context**
- [WIP branch](https://github.com/renovatebot/renovate/compare/master...TheKevJames:add-regex-versioner) (still has a few TODOs that I need to finish up or ask some questions about)
- [successfully works for the docker:python test case](https://github.com/TheKevJames/renovate-test/pull/34)

### PR DESCRIPTION
**Pull request by TheKevJames**
This branch provides an implementation for #4264 -- see discussion there.

I have a few TODOs still in this PR, but I'm starting to run into the limits of my understanding of TypeScript, JavaScript, and/or this repo. If anyone could help answer any of the open comments I've added on this PR, that would be fantastic!

---

Closes renovatebot#4264

### CODE REVIEW

File: lib/versioning/regex/index.ts
Comment (TheKevJames) [inline lib/versioning/regex/index.ts]: @ anyone-who-knows-javascript: Is there a way to follow this lint rule without breaking my usage of `this`?

@ maintainers: if not, can this rule be disabled?

File: lib/versioning/regex/index.ts
Comment (TheKevJames) [inline lib/versioning/regex/index.ts]: @ anyone-who-knows-javascript: Anyone know of a better way to get a list of the named capture groups in a `RegExp`? My current best-effort thought is below, but its pretty kludgey.

Running the regex against a successful string will let you introspect the object and get out the keys, but that requires knowledge of a working match:
```
> Object.keys("test".match(/(?<foo>te)(?<bar>st)/).groups)
(2) ["foo", "bar"]
```

File: lib/versioning/regex/index.ts
Comment (TheKevJames) [inline lib/versioning/regex/index.ts]: @ maintainers: how much validation should I be adding here?

File: lib/versioning/regex/index.ts
Comment (TheKevJames) [inline lib/versioning/regex/index.ts]: @ contributors: how should the various `VersioningApi` methods return if a version returned as a possible update is completely invalid (say, if someone releases a version `foo` to an otherwise-semver package)? I assume `isValid() == false` would be enough here, but which of the `VersioningApi` methods do I need to implement that for? Can I assume if `isValid()` is false then none of the other methods will get called?

File: lib/versioning/regex/index.ts
Comment (TheKevJames) [inline lib/versioning/regex/index.ts]: @ contributors: could I get a rundown of the difference between `isSingleVersion`, `isVersion`, and `isValid`? My impression is that for this versioner they'd all be the same (eg. any version that can be parsed as a version is inherently valid and not a range), but I got a bit lost following all the code paths here.

File: lib/versioning/regex/index.ts
Comment (ViceIce) [reply lib/versioning/regex/index.ts]: move `config` to module scope variable, then you dont have to use `this`. There can only be one instance of a versioning scheme.

File: lib/versioning/index.ts
Comment (ViceIce) [inline lib/versioning/index.ts]: please make `versionConfig` optional: `versionConfig?: string`

File: lib/manager/nuget/extract.ts
Comment (ViceIce) [inline lib/manager/nuget/extract.ts]: Revert this, because `versionConfig` should be optional.

File: lib/util/package-rules.js
Comment (ViceIce) [inline lib/util/package-rules.js]: Please do not remove object destructuring. see [destructuring assignment](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment)

File: lib/versioning/common.ts
Comment (ViceIce) [inline lib/versioning/common.ts]: should be void: `configure?(config: string): void;`

File: lib/versioning/regex/index.ts
Comment (ViceIce) [reply lib/versioning/regex/index.ts]: Here the different version apis are not consistent. Some return `null` or `undefined` as `false`.
This should be fixed in a future version.

File: lib/workers/pr/changelog/source-github.js
Comment (ViceIce) [inline lib/workers/pr/changelog/source-github.js]: again, please do not remove object destructuring.

File: lib/workers/pr/pr-body.js
Comment (ViceIce) [inline lib/workers/pr/pr-body.js]: object destructuring

File: lib/workers/repository/init/vulnerability.js
Comment (ViceIce) [inline lib/workers/repository/init/vulnerability.js]: revert, see other comments

File: lib/workers/repository/process/lookup/filter.js
Comment (ViceIce) [inline lib/workers/repository/process/lookup/filter.js]: object destructuring

File: lib/datasource/index.js
Comment (ViceIce) [inline lib/datasource/index.js]: object destructuring, see below

**Review by viceice — CHANGES_REQUESTED** [review]
Please revert all removed object destructuring assignments.

**Comment by rarkins** [comment]
Let's avoid any unnecessary refactoring in this PR. And for anything that's necessary, create and merge it as a separate PR first so that we can isolate just the new functionality here.

**Comment by rarkins** [comment]
@TheKevJames Configuration-wise, would there be any problem with including the regex string within the versionScheme itself? e.g. `"versionScheme": "regex:aaaaaaaaaaaaaaaaaaaaaa"`. Internally we'd need to do some extra work to split/massage the params but wouldn't it make the config cleaner?

File: lib/versioning/regex/index.ts
Comment (TheKevJames) [reply lib/versioning/regex/index.ts]: Ah perfect, that did the trick. Not sure why I thought I needed this in the first place -- thanks!

**Comment by TheKevJames** [comment]
@rarkins I could go either way on the configuration change -- personally, I find keeping `versionScheme` and `versionConfig` separate makes sense since they have different schemas and purposes, are parsed differently, and doing our own string parsing on top of that adds unnecessary constraints to scheme names. That said, I defer to the maintainers on the one, since it basically comes down to how y'all want the API.

Let me know if you'd prefer `versionScheme: "name(:config)?"`-style parsing and I'd be happy to switch over!

File: lib/versioning/regex/index.ts
Comment (TheKevJames) [reply lib/versioning/regex/index.ts]: Copying from a response from @rarkins (thanks!)

> `isSingleVersion` exists because versioning schemes have different syntax. e.g. for python you need `==1.0.0` while for other managers `1.0.0` means "1.0.0 or above", etc. So it returns true if the provided string represents a pinned/single version
>
> in this regex case I would not support ranges so isSingleVersion/isVersion/isValid would probably all return the same answer every time
>
> `isVersion` means the string is a version (e.g. `1.0.0`) while isValid can include ranges too (e.g. >= `1.0.0`)
> 
> `isValid` should return false for something like `foo`. Ideally your `isVersion` and `isSingleVersion` functions call to `isValid` first though perhaps, in order to be defensive
> 
> ideally the other functions would not get called by the application layer if `isValid` is false, but while writing the versioning API we can't be sure of that

**Comment by TheKevJames** [comment]
Huh, everything seems to have passed tests on Azure, not sure why they're reporting failures in GH...

File: lib/versioning/regex/index.ts
Comment (ViceIce) [inline lib/versioning/regex/index.ts]: @rarkins should we check for possible null refs?

**Comment by ViceIce** [comment]
Coverage is missing, so please add tests

https://dev.azure.com/renovatebot/6ed21661-b561-4777-8106-ca95b548903d/_apis/test/CodeCoverage/browse/244857/Code%20Coverage%20Report_65/lib.versioning.regex_ts.htm
https://dev.azure.com/renovatebot/6ed21661-b561-4777-8106-ca95b548903d/_apis/test/CodeCoverage/browse/244857/Code%20Coverage%20Report_65/lib.versioning_ts2.htm

File: lib/versioning/regex/index.ts
Comment (TheKevJames) [reply lib/versioning/regex/index.ts]: I was looking into this a while back, especially in response to https://github.com/renovatebot/renovate/pull/4273#discussion_r312353683, and went looking at the other schemes to see how its been done before -- as far as I could tell, most of the other schemes currently obey this format, where generally most `VersioningApi` methods will raise an exception on un-parseable version schemes (in anything but the `isValid` method(s)). Figured it might be best to keep with the current style, for now?

**Comment by rarkins** [comment]
I'm concerned about what happens if a repository has more than one regex `versionConfig`. Because these version schemes aren't Class-based, their "config" will be shared and hence non-deterministic because they could overwrite each other if executed async.

I have an idea, would like @ViceIce in particular to discuss it as you're strongest with TS/JS:

- Any versioning scheme that needs "configuration" (starting with `regex`) must be implemented as a Class
- `versioning.get()` should be updated to detect when Class-based is necessary

Because of this I'm flipping back to my original preference which was defining the config as part of the `versionScheme`. This approach makes it impossible to define regex without a config, and also keeps the existing versioning.get() signature the same.

**Comment by ViceIce** [comment]
:thinking:

1. Simplest would be to append the config as tring to the `versionScheme` so we have only string.

2. Or we can overload the get

```ts
get(scheme: string | VersionScheme): VersioningApi;
```

```ts
interface VersionScheme extends Record<string,any> {
  type: string;
}
```

and we need some instance, where the config can be stored.

Then we have to make sure, that we do not use object destruction on that api instance, otherwise we lose the `this` context.
The other way is to bind all api function to the config object, but that would be bad code style.

:thinking:

I think the best would be to use a real class for the versioning api, which will be initialized in `get` and don't use object destruction on that. This would change a lot of code. Maybe we should do that refactoring before we add the new `regex` scheme.

**Comment by rarkins** [comment]
> I think the best would be to use a real class for the versioning api, which will be initialized in get and don't use object destruction on that.

Wouldn't it be possible to keep existing versioning interfaces as-is and return a class only for the new regex one? As long as the function names are all the same, is it possible?

> Maybe we should do that refactoring before we add the new regex scheme.

Agreed

**Comment by ViceIce** [comment]
We can use the existing api, but we can't use object destruction on that api

```ts
 const { isValid } = get({scheme: 'regex', config: '...' });
  isValid(version) // error wrong this context!
```

```ts
 const versioning = get({scheme: 'regex', config: '...' });
  versioning.isValid(version) // ok versioning is this context
```

**Comment by rarkins** [comment]
> We can use the existing api, but we can't use object destruction on that api

I was definitely not aware of that. Any chance you can describe why? e.g. is it because class functions can't be destructured?

**Comment by ViceIce** [comment]
yes, it is a fundamental javascript design.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/this

**Comment by rarkins** [comment]
So to refactor in advance we need to make sure all `versioning.get()` calls are not destructured? Can they be destructured in the line immediately after or is that also not possible?

**Comment by ViceIce** [comment]
You can, but then you have to use `call` or `apply` 

```ts
isValid.call(versioning, version)
```

This is also bad code.

We have to think object orientated, because we have an instance with instance properties.

**Comment by TheKevJames** [comment]
@rarkins @ViceIce this conversation is definitely starting to go over my head in terms of the internals of how we want this library to work. Sounds like its _not_ the case that there can only be [one instance of a `versioningScheme`](https://github.com/renovatebot/renovate/pull/4273#discussion_r312351484)?

I can verify that avoiding destructuring allows the regex versioning API to be able to make use of `this` (since that was my original design before code review began). Additionally, its worth pointing out that this approach worked without needing to modify any of the other version schemes, eg. we can have only the complex ones become classes, if we'd like, and leave the others as-is.

**Comment by ViceIce** [comment]
@rarkins We don't have to change all schemes, but we should handle them as they where objects. So we can change one later if required.

@TheKevJames Sorry for the complex discussion. Because of many async calls it can happen that  calls have a different order. And then the `isVersion` call can be mixed with the `configure` calls

**Comment by ViceIce** [comment]
@TheKevJames can you add a new pr where you eliminate all the object destruction of the versioning api's. We can add in the `get` function somthing like

```ts
 return { ...api, config };
```

Then you can use it with `this` and it will override the default config.

@rarkins  What dou you think?

**Comment by rarkins** [comment]
@ViceIce unfortunately I'm a little bit lost in the JS complexity now too - I have stayed away from truly understanding `this` until now. In particular I don't understand the `return { ...api, config };` part of your suggestion.

My ideal is:
- Existing versioning schemes can stay as-is, namely shared/class-less/etc
- The new regex scheme needs to return a separate instance for each `versioning.get()` in order to allow use of multiple version configs in the same repo
- Ideally the downstream functions that use `versioning.get()` don't have to change any signatures

Is that in line with what you are suggesting?

**Comment by ViceIce** [comment]
Yes, that what i suggest. I think @TheKevJames can as first step add a pr which will remove the object destruction. The we can continue here.

**Comment by TheKevJames** [comment]
> Ideally the downstream functions that use versioning.get() don't have to change any signatures

I'm not sure this requirement is in-line with the requirement that we eliminate desctructuring of the `versioning.get()` return value -- my understanding from @ViceIce is that we _must_ make this change to support the rest?

**Comment by rarkins** [comment]
I'm working on that first PR now, I don't want to burn @TheKevJames out :)

**Comment by ViceIce** [comment]
Ok, was just a suggestion. 😄

**Comment by rarkins** [comment]
I've also refactored `versionConfig` back into `versionScheme`. Can anyone see any major downsides to this? I prefer it because then it won't really be possible to accidentally miss out a versionConfig for regex nor to include one for others. I guess the main thing we need to check for is someone just configuring `regex`?

**Comment by TheKevJames** [comment]
@rarkins I think the biggest downside I see is that it (at least personally) makes things a bit less clear: we still have to validate that eg. the `regex` scheme always has a `:<config>` and the other schemes don't, which is really no different than needing to validate that the `regex` scheme has a `versionScheme` and the others don't. I think it comes down to personal preference: we need to do the same validations in either case, so is it more clear to have a "scheme" and a "config", or is it more clear for the "scheme" to inherently include the config with some magic string parsing?

File: lib/versioning/regex/index.ts
Comment (ViceIce) [inline lib/versioning/regex/index.ts]:
```ts
let config: RegExp = null;
```

File: lib/versioning/index.ts
Comment (ViceIce) [inline lib/versioning/index.ts]:
```ts
let schemeName: string;
let schemeConfig: string;
```

File: lib/versioning/index.ts
Comment (ViceIce) [inline lib/versioning/index.ts]: New Idea: `configure` returns an `VersioningApi` object and we use that return value if `configure` exists.

So the Scheme can decide what to do with the config and how the object is constructed.

File: lib/versioning/regex/index.ts
Comment (ViceIce) [inline lib/versioning/regex/index.ts]: here we could return
```ts
return {...api, config: new RegExp(new_config)};
```

We should define an  interface for that

```ts
export interface RegExpVersionApi extends VersionApi {
  config: RegExp;
}
```

File: lib/versioning/regex/index.ts
Comment (ViceIce) [inline lib/versioning/regex/index.ts]: from my idea:
```ts
function parse(this: RegExpVersionApi, version: string): Version | null {
  const match = version.match(this.config);
```

https://github.com/Microsoft/TypeScript/wiki/What's-new-in-TypeScript#specifying-the-type-of-this-for-functions

File: lib/versioning/regex/index.ts
Comment (ViceIce) [inline lib/versioning/regex/index.ts]: for all parse we have to use `this.parse`. You can declare the `this` type like on `parse` function.

**Review by viceice — CHANGES_REQUESTED** [review]
@rarkins What do you think about my idea?

**Comment by rarkins** [comment]
@ViceIce it sounds interesting. Do you have any spare time for the refactoring? I don't think I'll get it right.

File: lib/versioning/regex/index.ts
Comment (rarkins) [reply lib/versioning/regex/index.ts]: @ViceIce btw have you seen this new feature?

image

Whenever the suggestion is just a straight code change, it's then one click to apply

Comment (rarkins) [reply lib/versioning/regex/index.ts]: e.g.

\`\`\`suggestion
let config: RegExp = null;
\`\`\`

Comment (rarkins) [reply lib/versioning/regex/index.ts]: And then the result:
image

Comment (ViceIce) [reply lib/versioning/regex/index.ts]: Nice, will try it

File: lib/versioning/index.ts
Comment (rarkins) [inline lib/versioning/index.ts]:
```suggestion
  let schemeName: string;
```

File: lib/versioning/index.ts
Comment (rarkins) [inline lib/versioning/index.ts]:
```suggestion
  let schemeConfig: string;
```

File: lib/versioning/regex/index.ts
Comment (ViceIce) [inline lib/versioning/regex/index.ts]:
```suggestion
import safe from 'safe-regex';
```

**Comment by ViceIce** [comment]
@rarkins  see https://github.com/TheKevJames/renovate/pull/2 for my proposed chages

**Comment by ViceIce** [comment]
`safe-regex` is not working here

**Comment by rarkins** [comment]
Apart from fixing `safe-regex` use, what's remaining to be done for this PR?

**Comment by ViceIce** [comment]
we have to merge my pr, rebase and fix conflicts

**Comment by ViceIce** [comment]
@rarkins I can rebase my changes to current master and open a new superceeding pr

**Comment by rarkins** [comment]
@ViceIce 	ok

**Pull request closed without merging**
